src/core/thread/mle.cpp

```diff
@@ -917,9 +925,13 @@ otError Mle::SetMeshLocalPrefix(const otMeshLocalPrefix &aMeshLocalPrefix)
     if (mRole == OT_DEVICE_ROLE_LEADER)
     {
         netif.RemoveUnicastAddress(mLeaderAloc);
-        memcpy(mLeaderAloc.GetAddress().mFields.m8, mMeshLocal64.GetAddress().mFields.m8, 8);
+        memcpy(mLeaderAloc.GetAddress().mFields.m8, aMeshLocalPrefix.m8, sizeof(aMeshLocalPrefix));
         netif.AddUnicastAddress(mLeaderAloc);
     }
+    else
+    {
+        memcpy(mLeaderAloc.GetAddress().mFields.m8, aMeshLocalPrefix.m8, sizeof(aMeshLocalPrefix));
```

Comment:
`mLeaderAloc` may only be significant for FTD device. We may further enhance it in the further.
How about deferring to update its prefix when it is to be added in `Mle::AddLeaderAloc()`?